### ISSUE TITLE
Skip() and Take() produce the wrong results when querying entities with a many-to-many relationship.

### ISSUE THREAD

**Issue by maxmantz** [aspnet/EntityFramework#3820]
I have encountered a bug in EF7 (beta8) where using `Skip()` & `Take()` produce the wrong results when trying to query entites with a many-to-many relationship. See [my repo](https://github.com/maxmantz/EF7-ManyToManyBug) for details.

EDIT: the bug doesn't appear when using the in memory database.

**Comment by Muchiachio** [comment]
Maybe it's related, using `Skip(0).Take(0)` in rc1-final seems to always result in
`The number of rows provided for a FETCH clause must be greater then zero.`

**Comment by maxmantz** [comment]
@Muchiachio I don't think so. Using `Skip(0).Take(0)` will always provide an empty result set, so the resulting warning makes perfect sense.

**Comment by Muchiachio** [comment]
Yes, but it throws exception, on first enumeration, but it shouldn't.

**Comment by mikary** [comment]
This issue appears to be the same as #4083 which was resolved by 54b5c68

Given the provided query
```cs
            var students = await this.ctx.Students
                .Where(s => true)
                .Include(s => s.StudentCourses)
                .ThenInclude(sc => sc.Course)
                .Skip(offset)
                .Take(limit)
                .ToListAsync();
```

The current dev version of EF generates the following SQL:
```sql
SELECT [s].[Id], [s].[Name]
FROM [Student] AS [s]
WHERE 1 = 1
ORDER BY [s].[Id]
OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY


SELECT [s].[StudentId], [s].[CourseId], [c].[Id], [c].[Name]
FROM [StudentCourse] AS [s]
INNER JOIN (
    SELECT DISTINCT [t0].*
    FROM (
        SELECT [s].[Id]
        FROM [Student] AS [s]
        WHERE 1 = 1
        ORDER BY [s].[Id]
        OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY
    ) AS [t0]
) AS [s0] ON [s].[StudentId] = [s0].[Id]
INNER JOIN [Course] AS [c] ON [s].[CourseId] = [c].[Id]
ORDER BY [s0].[Id]
```

In the original issue, the query does not include the final `ORDER BY` which results in materialization issues like the one described.

To be sure this issue is resolved, I ran the following (Note: the additional `OrderBy` is necessary for `Skip` and `Take` on SqlServer because the default ordering is by `Customer.Id`):
```cs
    public class Program
    {
        public static void Main(string[] args)
        {
            var inMemoryStudents = new List<Student>();
            var inMemoryCourses = new List<Course>();
            var inMemoryStudentCourses = new List<StudentCourse>();

            using (var context = new DemoContext())
            {
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();

                for (int i = 1; i <= 100; i++)
                {
                    var student = new Student { Name = $"Student {i}" };
                    var course = new Course { Name = $"Course {i}" };

                    context.Students.Add(student);
                    context.Courses.Add(course);

                    inMemoryStudents.Add(new Student { Name = $"Student {i}" });
                    inMemoryCourses.Add(new Course { Name = $"Course {i}" });
                }

                context.SaveChanges();

                var students = context.Students.ToList();

                foreach (var student in students)
                {
                    var inMemoryStudent = inMemoryStudents.Single(s => s.Name == student.Name);

                    for (int i = 1; i <= 50; i++)
                    {
                        var course = context.Courses.Where(c => c.Name.Equals($"Course {i}")).FirstOrDefault();
                        var inMemoryCourse = inMemoryCourses.Single(c => c.Name == course.Name);

                        if (course != null)
                        {
                            var studentCourse = new StudentCourse(student, course);
                            context.StudentCourses.Add(studentCourse);

                            var inMemoryStudentCourse = new StudentCourse(inMemoryStudent, inMemoryCourse);
                            inMemoryStudentCourses.Add(inMemoryStudentCourse);
                            inMemoryStudent.StudentCourses.Add(inMemoryStudentCourse);
                            inMemoryCourse.StudentCourses.Add(inMemoryStudentCourse);
                        }
                    }
                }
                context.SaveChanges();
            }

            try
            {
                for (var offset = 0; offset < 101; offset += 10)
                {
                    for (var limit = 1; limit <= 100; limit++)
                    {
                        Console.WriteLine("Offset: {0} Limit: {1}", offset, limit);

                        using (var context = new DemoContext())
                        {
                            var expected = inMemoryStudents
                                .OrderBy(s => s.Name)
                                .Skip(offset)
                                .Take(limit)
                                .ToList();

                            var actual = context.Students
                                .Where(s => true)
                                .Include(s => s.StudentCourses)
                                .ThenInclude(sc => sc.Course)
                                .OrderBy(s => s.Name)
                                .Skip(offset)
                                .Take(limit)
                                .ToList();

                            if (expected.Count != actual.Count)
                            {
                                throw new Exception("Incorrect number of Students");
                            }

                            foreach (var expectedStudent in expected)
                            {
                                var actualStudent = actual.SingleOrDefault(s => s.Name == expectedStudent.Name);

                                if(actualStudent == null)
                                {
                                    throw new Exception($"Student not found - {expectedStudent.Name}");
                                }


                                if (expectedStudent.StudentCourses.Count != actualStudent.StudentCourses.Count)
                                {
                                    throw new Exception("Incorrect number of StudentCourses");
                                }

                                foreach (var expectedStudentCourse in expectedStudent.StudentCourses)
                                {
                                    var actualStudentCourse = actualStudent.StudentCourses
                                        .Single(sc => (sc.Course.Name == expectedStudentCourse.Course.Name)
                                            && (sc.Student.Name == expectedStudentCourse.Student.Name));

                                    if(actualStudentCourse == null)
                                    {
                                        throw new Exception($"StudentCourse not found - Student: {expectedStudentCourse.Student.Name} Course: {actualStudentCourse.Course.Name}");
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed: {e.Message}");
            }

            Console.WriteLine("Done.");
            Console.ReadLine();
        }
    }

    public class DemoContext : DbContext
    {
        public DbSet<Course> Courses { get; set; }

        public DbSet<Student> Students { get; set; }

        public DbSet<StudentCourse> StudentCourses { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Student>().HasMany(e => e.StudentCourses);
            modelBuilder.Entity<Course>().HasMany(e => e.StudentCourses);
            modelBuilder.Entity<StudentCourse>().HasKey(x => new { x.StudentId, x.CourseId });
            base.OnModelCreating(modelBuilder);
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=ManyToManyDemo;Trusted_Connection=True;");
            //optionsBuilder.UseInMemoryDatabase();
        }
    }

    public class Course
    {
        public int Id { get; set; }

        public string Name { get; set; }

        public ICollection<StudentCourse> StudentCourses { get; set; } = new List<StudentCourse>();
    }

    public class Student
    {
        public int Id { get; set; }

        public string Name { get; set; }

        public ICollection<StudentCourse> StudentCourses { get; set; } = new List<StudentCourse>();
    }

    public class StudentCourse
    {
        public int StudentId { get; set; }

        public Student Student { get; set; }

        public int CourseId { get; set; }

        public Course Course { get; set; }

        public StudentCourse()
        {
        }

        public StudentCourse(Student student, Course course)
        {
            StudentId = student.Id;
            Student = student;
            CourseId = course.Id;
            Course = course;
        }
    }
```
